tools/smartdnsblock/smartdnsblock/smartdnsblock.cpp

```diff
@@ -0,0 +1,135 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// If this doesn't come first, there will be compile errors.
+#include <winsock2.h>
+
+#include <iostream>
+
+#include <iphlpapi.h>
+
+#include <fwpmtypes.h>
+#include <fwpmu.h>
+
+#pragma comment(lib, "iphlpapi.lib")
+#pragma comment(lib, "fwpuclnt.lib")
+
+using namespace ::std;
+
+PCWSTR TAP_DEVICE_NAME = L"outline-tap0";
+ULONG GET_ADAPTERS_ADDRESSES_BUFFER_SIZE = 16384;
+
+PCWSTR FILTER_PROVIDER_NAME = L"Outline";
+
+int main(int argc, char **argv) {
+  // Lookup the interface index of outline-tap0.
+  PIP_ADAPTER_ADDRESSES adaptersAddresses =
+      (IP_ADAPTER_ADDRESSES *)malloc(GET_ADAPTERS_ADDRESSES_BUFFER_SIZE);
+  DWORD result = GetAdaptersAddresses(AF_INET, 0, NULL, adaptersAddresses,
+                                      &GET_ADAPTERS_ADDRESSES_BUFFER_SIZE);
+  if (result != NO_ERROR) {
+    cerr << "could not fetch network device list: " << result << endl;
+    return 1;
+  }
+
+  UINT32 interfaceIndex;
+  PIP_ADAPTER_ADDRESSES adapterAddress = adaptersAddresses;
+  while (adapterAddress && wcscmp(TAP_DEVICE_NAME, adapterAddress->FriendlyName) != 0) {
+    adapterAddress = adapterAddress->Next;
+  }
+
+  if (!adapterAddress) {
+    wcerr << "could not find " << TAP_DEVICE_NAME << endl;
+    return 1;
+  }
+
+  interfaceIndex = adapterAddress->IfIndex;
+  wcout << "found " << TAP_DEVICE_NAME << " at index " << interfaceIndex << endl;
+
+  // Connect to the filtering engine. By using a dynamic session, all of our changes are
+  // *non-destructive* and will vanish on exit/crash/whatever.
+  FWPM_SESSION0 session;
+  memset(&session, 0, sizeof(session));
+  session.flags = FWPM_SESSION_FLAG_DYNAMIC;
+
+  HANDLE engine = 0;
+  result = FwpmEngineOpen0(NULL, RPC_C_AUTHN_DEFAULT, NULL, &session, &engine);
+  if (result != ERROR_SUCCESS) {
+    cerr << "could not connect to to filtering engine: " << result << endl;
+    return 1;
+  }
+  cout << "connected to filtering engine" << endl;
+
+  // Create our filters.
+  //
+  // We create two filters, each with multiple conditions:
+  //  - The first filter blocks all UDP traffic on port 53.
+  //  - The second "extends" the first filter to allow such traffic *on the TAP device*.
+  //
+  // This approach of "layering" conditions is the same as that used in the SDK documentation:
+  //   https://docs.microsoft.com/en-us/windows/desktop/fwp/reserving-ports
+  //
+  // Note:
+  //  - Since OutlineService adds a blanket block on all IPv6 traffic, we only need to create IPv4
+  //    filters.
+  //  - Thanks to the simplicity of the filters and how they will be automatically destroyed on
+  //    exit, there's no need to use a transaction here.
```

Comment:
Do we know why this happens? Is it documented somewhere?

Reply:
Do you mean why is a transaction not required? Trancsactions aren't **required** in WFP - if in use they will help you clean up complex rules. Since our rules are simple and transient, I just figured there's no need.

Reply:
Sorry, I wasn't clear. I meant to ask if you know why the rules are automatically removed once the process exits.

Reply:
Oh, because smartdnsblock sets the `FWPM_SESSION_FLAG_DYNAMIC` session flag:
https://docs.microsoft.com/en-us/windows/desktop/api/fwpmtypes/ns-fwpmtypes-fwpm_session0_